xrspatial/experimental/polygonize.py

```diff
@@ -0,0 +1,590 @@
+            # Follow hole of polygon on N side of pixel ij-nx facing W.
+            region, points = _follow(regions, visited, nx, ny, ij-nx, True)
+            if transform is not None:
+                _transform_points(points, transform)
+            # Polygon index is one less than region as region 0 has no
+            # polygons.
+            polygons[region-1].append(points)
+
+    return column, polygons
+
+
+def _to_awkward(
+    column: List[Union[int, float]],
+    polygon_points: List[np.ndarray],
+):
+    import awkward as ak
+    ak_array = ak.Array(polygon_points)
+    return column, ak_array
+
+
+def _to_geopandas(
+    column: List[Union[int, float]],
+    polygon_points: List[np.ndarray],
+    column_name: str,
+):
+    import geopandas as gpd
+    from shapely.geometry import Polygon
+
+    # Convert list of point arrays to shapely Polygons.
+    polygons = list(map(
+        lambda points: Polygon(points[0], points[1:]), polygon_points))
+
+    df = gpd.GeoDataFrame({column_name: column, "geometry": polygons})
+    return df
+
+
+def _to_spatialpandas(
+    column: List[Union[int, float]],
+    polygon_points: List[np.ndarray],
+    column_name: str,
+):
+    from spatialpandas import GeoDataFrame
+    from spatialpandas.geometry import PolygonArray
+
+    # spatialpandas expects 1d numpy arrays.
+    for i, arrays in enumerate(polygon_points):
+        polygon_points[i] = \
+            list(map(lambda array: np.reshape(array, -1), arrays))
+
+    df = GeoDataFrame({
+        column_name: column, "geometry": PolygonArray(polygon_points)})
+    return df
+
+
+def _polygonize_numpy(
+    values: np.ndarray,
+    mask: Optional[np.ndarray],
+    connectivity_8: bool,
+    transform: Optional[np.ndarray],
+    column_name: str,
+    return_type: str,
+):
+    ny, nx = values.shape
+    if nx == 1:
+        # Algorithm requires nx > 1 to differentiate between facing E
+        # (forward == 1) and facing N (forward == nx), so add extra column to
+        # values array and mask the column out.
+        nx = 2
+        values = np.hstack((values, np.empty_like(values)))
+        if mask is not None:
+            mask = np.hstack((mask, np.zeros_like(mask)))
+        else:
+            mask = np.zeros_like(values, dtype=bool)
+            mask[:, 0] = True
+
+    values = values.ravel()
+    if mask is not None:
+        mask = mask.ravel()
+
+    column, polygon_points = _scan(
+        values, mask, connectivity_8, transform, nx, ny)
+
+    if return_type == "numpy":
+        return column, polygon_points
+    elif return_type == "awkward":
+        return _to_awkward(column, polygon_points)
+    elif return_type == "geopandas":
+        return _to_geopandas(column, polygon_points, column_name)
+    elif return_type == "spatialpandas":
+        return _to_spatialpandas(column, polygon_points, column_name)
+    else:
+        raise ValueError(f"Invalid return_type '{return_type}'")
```

Comment:
It would be better to validate `return_type` in `polygonize()` IMO

Reply:
OK, will do. I'll also add a new test passing an invalid string for `return_type` and checking that an appropriate exception is raised.